Fix multiprocessing start method for Python 3.14 compatibility (Issue #169252)

ada-ggf25 opened the pull request:
## Description

This PR fixes compatibility issues with Python 3.14's change to the default multiprocessing start method. In Python 3.14, the default start method on Unix platforms (excluding macOS) changes from `'fork'` to `'forkserver'`. Code that relies on `'fork'` being the default will break without explicit context specification.

Fixes #169252

## Problem

Several parts of PyTorch use `multiprocessing.Process()`, `multiprocessing.Pool()`, and `multiprocessing.Manager()` without explicitly setting the start method, relying on the system default. With Python 3.14, this default changes, which could cause:
- Test failures in tests that specifically test fork behaviour
- Issues with distributed debug server process creation
- Problems with ONNX Runtime process execution in isolated contexts

## Solution

This PR explicitly sets the multiprocessing start method in all affected locations:
- For code that specifically needs `'fork'` (e.g., fork behaviour tests), we explicitly request `'fork'`
- For code that doesn't require a specific method, we use `'spawn'` as it's the most compatible across platforms

## Changes Made

### 1. `torch/distributed/debug/__init__.py`
- Added explicit `'spawn'` context creation as module-level `_MP_CONTEXT`
- Updated `multiprocessing.Process()` to use `_MP_CONTEXT.Process()`
- Added explanatory comments about Python 3.14 compatibility

**Rationale**: The debug server doesn't specifically need `'fork'`, so `'spawn'` is a safe choice that works on all 
platforms and is recommended for distributed training scenarios.

### 2. `test/test_cuda.py` (test_is_pinned_no_context)
- Added explicit `'fork'` context using `multiprocessing.get_context('fork')`
- Updated `multiprocessing.Pipe()` to use `mp_ctx.Pipe()`
- Updated `multiprocessing.Process()` to use `mp_ctx.Process()`
- Added explanatory comments about Python 3.14 compatibility

**Rationale**: This test explicitly tests fork behaviour (as indicated by the function name `fork_and_check_is_pinned`), so it needs `'fork'` explicitly.

### 3. `test/onnx/torchlib/ops_test_common.py` (_safe_ort_session_run)
- Added explicit `'spawn'` context using `multiprocessing.get_context('spawn')`
- Updated `multiprocessing.Manager()` to use `mp_ctx.Manager()`
- Updated `multiprocessing.Process()` to use `mp_ctx.Process()`
- Added explanatory comments about Python 3.14 compatibility

**Rationale**: ONNX Runtime tests don't specifically need `'fork'`, and `'spawn'` is more compatible with external libraries like ONNX Runtime.

## Testing

-  All changes follow PyTorch coding conventions
-  No linting errors
- Changes are minimal and focused

## Additional Notes

- This change maintains backward compatibility with Python < 3.14

ada-ggf25 committed:
fix(test): explicitly use fork context in test_is_pinned_no_context for Python 3.14 compatibility

The test_is_pinned_no_context test in test_cuda.py explicitly tests fork
behaviour but was relying on the default multiprocessing start method being
'fork'. In Python 3.14, the default start method changes from 'fork' to
'forkserver' on Unix platforms (excluding macOS), which would cause this test
to fail.

This change explicitly requests the 'fork' context using
multiprocessing.get_context('fork') to ensure the test continues to work
correctly with Python 3.14 whilst maintaining the intended fork behaviour
testing.

Changes:
- Added explicit 'fork' context creation in fork_and_check_is_pinned()
- Updated multiprocessing.Pipe() to use mp_ctx.Pipe()
- Updated multiprocessing.Process() to use mp_ctx.Process()
- Added explanatory comments about Python 3.14 compatibility

Fixes #169252

ada-ggf25 committed:
fix(test): explicitly use spawn context in _safe_ort_session_run for Python 3.14 compatibility

The _safe_ort_session_run function in test/onnx/torchlib/ops_test_common.py
uses multiprocessing.Manager() and multiprocessing.Process() without explicitly
setting the start method, relying on the default being 'fork'. In Python 3.14,
the default start method changes from 'fork' to 'forkserver' on Unix platforms
(excluding macOS), which could cause issues with ONNX Runtime process execution.

This change explicitly requests the 'spawn' context using
multiprocessing.get_context('spawn') to ensure the function continues to work
correctly with Python 3.14. The 'spawn' method is chosen as it's the most
compatible across platforms and doesn't have the same limitations as 'fork'
when dealing with external libraries like ONNX Runtime.

Changes:
- Added explicit 'spawn' context creation in _safe_ort_session_run()
- Updated multiprocessing.Manager() to use mp_ctx.Manager()
- Updated multiprocessing.Process() to use mp_ctx.Process()
- Added explanatory comments about Python 3.14 compatibility

Fixes #169252

ada-ggf25 committed:
fix(distributed): explicitly use spawn context in debug server for Python 3.14 compatibility

The start_debug_server function in torch/distributed/debug/__init__.py uses
multiprocessing.Process() without explicitly setting the start method, relying
on the default being 'fork'. In Python 3.14, the default start method changes
from 'fork' to 'forkserver' on Unix platforms (excluding macOS), which could
cause issues with the distributed debug server process creation.

This change explicitly requests the 'spawn' context using
multiprocessing.get_context('spawn') to ensure the debug server continues to
work correctly with Python 3.14. The 'spawn' method is chosen as it's the most
compatible across platforms and is the recommended method for distributed
training scenarios.

Changes:
- Added explicit 'spawn' context creation as module-level _MP_CONTEXT
- Updated multiprocessing.Process() to use _MP_CONTEXT.Process()
- Added explanatory comments about Python 3.14 compatibility

Fixes #169252

ada-ggf25 committed:
refactor(distributed): simplify Process instantiation in start_debug_server

The start_debug_server function in torch/distributed/debug/__init__.py has been updated to streamline the instantiation of the multiprocessing.Process. The changes remove unnecessary line breaks for improved readability while maintaining the existing functionality.

Changes:
- Simplified the instantiation of _DEBUG_SERVER_PROC by combining arguments into a single line.

No functional changes were made; this is purely a code readability improvement.

ada-ggf25 committed:
Merge branch 'pytorch:main' into Issue_169252

divyanshk reviewed (changes requested):
I think these tests are compatible with forkserver and do not need updates.

ada-ggf25 committed:
test(cuda): simplify multiprocessing context in fork_and_check_is_pinned

Remove explicit 'fork' context creation and revert to using default
multiprocessing context. This simplifies the code by removing the
workaround for Python 3.14 compatibility that explicitly set the
context to 'fork'.

Changes:
- Replace mp_ctx.Pipe() with multiprocessing.Pipe()
- Replace mp_ctx.Process() with multiprocessing.Process()
- Remove Python 3.14 compatibility comments and context setup

ada-ggf25 committed:
refactor(distributed): simplify multiprocessing context in debug server

Remove explicit 'spawn' context creation and revert to using default
multiprocessing context. This simplifies the code by removing the
workaround for Python 3.14 compatibility that explicitly set the
context to 'spawn'.

Changes:
- Remove _MP_CONTEXT variable and context setup
- Replace _MP_CONTEXT.Process() with multiprocessing.Process()
- Remove Python 3.14 compatibility comments

ada-ggf25 committed:
test(onnx): simplify multiprocessing context in _safe_ort_session_run

Remove explicit 'spawn' context creation and revert to using default
multiprocessing context. This simplifies the code by removing the
workaround for Python 3.14 compatibility that explicitly set the
context to 'spawn'.

Changes:
- Remove mp_ctx variable and context setup
- Replace mp_ctx.Manager() with multiprocessing.Manager()
- Replace mp_ctx.Process() with multiprocessing.Process()
- Remove Python 3.14 compatibility comments

divyanshk commented on torch/distributed/debug/__init__.py:
```diff
@@ -18,6 +18,11 @@
+# This ensures compatibility with Python 3.14 where default changes from 'fork' to 'forkserver'
+# We use 'spawn' here as it's the most compatible across platforms
+_MP_CONTEXT = multiprocessing.get_context("spawn")
+
```
Thanks @ada-ggf25 Closing the PR would be good.
